Refactor to reduce the amount of unsafe and duplicated code.

Diggsey opened the pull request:
- Replace uses of `pin_utils` with the safer `pin_project` crate
- Implement macros to allow building new combinators from existing ones more
  easily
- Add an internal `fns` module as a workaround until `impl Trait` is usable in more places.
- Get rid of Chain as the building block for future combinators, and instead
  build from `Map` and `Flatten` primitives
- Delete a lot of code which is no longer necessary.

There should not be any breaking changes as I have endeavoured to keep all type signatures exactly the same.

cramertj commented:
Wow, at a first glance this looks like a great improvement! Sadly though, this change is pretty large and exceeds my review capacity at the moment. I'll try to give it a read-through in the next week or so, but if either @Nemo157 or @taiki-e are available to take a look I'd greatly appreciate it :)

Diggsey commented:
Rebased to catch up with master.

taiki-e commented on futures-util/src/future/try_future/try_flatten_err.rs:
```diff
@@ -0,0 +1,61 @@
+                    self.set(TryFlattenErr::Empty);
+                    break output;
+                },
+                TryFlattenErr::Empty => return Poll::Pending,
```
```suggestion
                TryFlattenErr::Empty => panic!("TryFlattenErr polled after completion"),
```

Diggsey replied:
👍

taiki-e commented on futures-util/src/future/try_future/mod.rs:
````diff
@@ -449,7 +526,7 @@ pub trait TryFutureExt: TryFuture {
     /// # });
     /// ```
-    fn try_flatten_stream(self) -> TryFlattenStream<Self>
+    fn try_flatten_stream(self) -> TryFlattenStream<Self, Self::Ok>
````
Do you think we can avoid this change?

Diggsey replied:
Yep 👍 

One thing I found was that the code is quite inconsistent in when it uses type parameters vs extra bounds. (Ie. in this case, removing the extra type parameter means putting an extra bound on the `Fut` type).

It seems in some places it tries to avoid bounds on the types themselves (which does make them easier to use as you don't have to repeat those bounds everywhere) but in other places it prefers to reduce the number of type parameters.

Luckily, with the macro system it is very easy to switch back and forth.

taiki-e commented on futures-util/src/stream/stream/mod.rs:
````diff
@@ -550,7 +563,7 @@ pub trait StreamExt: Stream {
     /// # });
     /// ```
-    fn flatten(self) -> Flatten<Self>
+    fn flatten(self) -> Flatten<Self, Self::Item>
````
Same here

Diggsey replied:
👍

taiki-e commented on futures-util/src/future/maybe_done.rs:
```diff
@@ -48,21 +47,25 @@ impl<Fut: Future> MaybeDone<Fut> {
+        // state, which does not have any pinning guarantees to uphold.
+        //
+        // Hopefully `pin_project` will support this safely soon:
+        // https://github.com/taiki-e/pin-project/issues/184
```
cc https://github.com/taiki-e/pin-project/issues/184, in the public types, it is certainly not easy to use the way I said.

Diggsey replied:
Yeah this file is what lead me to that problem :)

Diggsey commented on futures-util/src/sink/with.rs:
```diff
@@ -135,16 +102,20 @@ impl<Si, Item, U, Fut, F, E> Sink<U> for With<Si, Item, U, Fut, F>
+        #[project]
+        let With { mut state, f, .. } = self.project();
+
+        assert!(state.is_none());
```
Cool, I will leave it in for now then.

Diggsey commented on futures-util/src/stream/stream/filter.rs:
```diff
@@ -103,34 +66,31 @@ impl<St, Fut, F> FusedStream for Filter<St, Fut, F>
+        let Filter { mut stream, f, mut pending_fut, pending_item } = self.project();
+        Poll::Ready(loop {
+            if let Some(fut) = pending_fut.as_mut().as_pin_mut() {
+                if ready!(fut.poll(cx)) {
```
👍

Diggsey commented on futures-util/src/stream/try_stream/try_filter.rs:
```diff
@@ -105,29 +69,26 @@ impl<St, Fut, F> Stream for TryFilter<St, Fut, F>
+        let TryFilter { mut stream, f, mut pending_fut, pending_item } = self.project();
+        Poll::Ready(loop {
+            if let Some(fut) = pending_fut.as_mut().as_pin_mut() {
+                if ready!(fut.poll(cx)) {
```
👍

Diggsey commented on futures-util/src/future/future/flatten.rs:
```diff
@@ -1,56 +1,165 @@
+                    self.set(Flatten::Empty);
+                    break output;
+                },
+                Flatten::Empty => return Poll::Pending,
```
Cool, in the original `Chain` based implementation it seems to erroneously use `unreachable!()`:

https://github.com/rust-lang/futures-rs/blob/master/futures-util/src/future/future/chain.rs#L50

So I wasn't quite sure what it should do here, but I will make it panic as you suggest.

Diggsey commented on futures-util/src/future/try_future/try_flatten.rs:
```diff
@@ -0,0 +1,180 @@
+                    self.set(TryFlatten::Empty);
+                    break output;
+                },
+                TryFlatten::Empty => return Poll::Pending,
```
👍

Diggsey commented:
Fixed an issue with the rebase and created a new commit with the changes as per review.

taiki-e commented on futures-util/src/stream/stream/filter.rs:
```diff
@@ -103,34 +66,32 @@ impl<St, Fut, F> FusedStream for Filter<St, Fut, F>
+        let Filter { mut stream, f, mut pending_fut, pending_item } = self.project();
+        Poll::Ready(loop {
+            if let Some(fut) = pending_fut.as_mut().as_pin_mut() {
+                if ready!(fut.poll(cx)) {
+                    pending_fut.set(None);
```
```suggestion
                let yield_item = ready!(fut.poll(cx));
                pending_fut.set(None);
                if yield_item {
```
I mean "if fut.poll returns Ready", not "if ready!(fut.poll) returns true" (I'm sorry if the comment was hard to understand)

taiki-e commented on futures-util/src/stream/try_stream/try_filter.rs:
```diff
@@ -105,29 +69,27 @@ impl<St, Fut, F> Stream for TryFilter<St, Fut, F>
+        let TryFilter { mut stream, f, mut pending_fut, pending_item } = self.project();
+        Poll::Ready(loop {
+            if let Some(fut) = pending_fut.as_mut().as_pin_mut() {
+                if ready!(fut.poll(cx)) {
+                    pending_fut.set(None);
```
```suggestion
                let yield_item = ready!(fut.poll(cx));
                pending_fut.set(None);
                if yield_item {
```

taiki-e commented on futures-util/src/future/future/mod.rs:
```diff
@@ -7,51 +7,85 @@ use super::{assert_future, Either};
+    /// Future for the [`map_into`](FutureExt::map_into) combinator.
+    MapInto<Fut, T>(
+        Map<Fut, IntoFn<T>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut| Map::new(x, into_fn())]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut| Map::new(x, into_fn())]
```

taiki-e commented on futures-util/src/future/future/mod.rs:
```diff
@@ -7,51 +7,85 @@ use super::{assert_future, Either};
+    /// Future for the [`then`](FutureExt::then) method.
+    Then<Fut1, Fut2, F>(
+        flatten::Flatten<Map<Fut1, F>, Fut2>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut1, y: F| flatten::Flatten::new(Map::new(x, y))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut1, y: F| flatten::Flatten::new(Map::new(x, y))]
```

taiki-e commented on futures-util/src/future/future/mod.rs:
```diff
@@ -7,51 +7,85 @@ use super::{assert_future, Either};
+    /// Future for the [`inspect`](FutureExt::inspect) method.
+    Inspect<Fut, F>(
+        map::Map<Fut, InspectFn<F>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut, f: F| map::Map::new(x, inspect_fn(f))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut, f: F| map::Map::new(x, inspect_fn(f))]
```

taiki-e commented on futures-util/src/future/future/mod.rs:
```diff
@@ -7,51 +7,85 @@ use super::{assert_future, Either};
+    /// Future for the [`never_error`](super::FutureExt::never_error) combinator.
+    NeverError<Fut>(
+        Map<Fut, OkFn<Never>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut| Map::new(x, ok_fn())]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut| Map::new(x, ok_fn())]
```

taiki-e commented on futures-util/src/future/future/mod.rs:
```diff
@@ -7,51 +7,85 @@ use super::{assert_future, Either};
+    /// Future for the [`unit_error`](super::FutureExt::unit_error) combinator.
+    UnitError<Fut>(
+        Map<Fut, OkFn<()>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut| Map::new(x, ok_fn())]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut| Map::new(x, ok_fn())]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`and_then`](TryFutureExt::and_then) method.
+    AndThen<Fut1, Fut2, F>(
+        TryFlatten<MapOk<Fut1, F>, Fut2>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut1, f: F| TryFlatten::new(MapOk::new(x, f))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut1, f: F| TryFlatten::new(MapOk::new(x, f))]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`or_else`](TryFutureExt::or_else) method.
+    OrElse<Fut1, Fut2, F>(
+        TryFlattenErr<MapErr<Fut1, F>, Fut2>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut1, f: F| TryFlattenErr::new(MapErr::new(x, f))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut1, f: F| TryFlattenErr::new(MapErr::new(x, f))]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`err_into`](TryFutureExt::err_into) method.
+    ErrInto<Fut, E>(
+        MapErr<Fut, IntoFn<E>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut| MapErr::new(x, into_fn())]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut| MapErr::new(x, into_fn())]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`ok_into`](TryFutureExt::ok_into) method.
+    OkInto<Fut, E>(
+        MapOk<Fut, IntoFn<E>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut| MapOk::new(x, into_fn())]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut| MapOk::new(x, into_fn())]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`inspect_ok`](super::TryFutureExt::inspect_ok) method.
+    InspectOk<Fut, F>(
+        Inspect<IntoFuture<Fut>, InspectOkFn<F>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut, f: F| Inspect::new(IntoFuture::new(x), inspect_ok_fn(f))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut, f: F| Inspect::new(IntoFuture::new(x), inspect_ok_fn(f))]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`inspect_err`](super::TryFutureExt::inspect_err) method.
+    InspectErr<Fut, F>(
+        Inspect<IntoFuture<Fut>, InspectErrFn<F>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut, f: F| Inspect::new(IntoFuture::new(x), inspect_err_fn(f))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut, f: F| Inspect::new(IntoFuture::new(x), inspect_err_fn(f))]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`map_ok`](TryFutureExt::map_ok) method.
+    MapOk<Fut, F>(
+        Map<IntoFuture<Fut>, MapOkFn<F>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut, f: F| Map::new(IntoFuture::new(x), map_ok_fn(f))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut, f: F| Map::new(IntoFuture::new(x), map_ok_fn(f))]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`map_err`](TryFutureExt::map_err) method.
+    MapErr<Fut, F>(
+        Map<IntoFuture<Fut>, MapErrFn<F>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut, f: F| Map::new(IntoFuture::new(x), map_err_fn(f))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut, f: F| Map::new(IntoFuture::new(x), map_err_fn(f))]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`map_ok_or_else`](TryFutureExt::map_ok_or_else) method.
+    MapOkOrElse<Fut, F, G>(
+        Map<IntoFuture<Fut>, MapOkOrElseFn<F, G>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut, f: F, g: G| Map::new(IntoFuture::new(x), map_ok_or_else_fn(f, g))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut, f: F, g: G| Map::new(IntoFuture::new(x), map_ok_or_else_fn(f, g))]
```

taiki-e commented on futures-util/src/future/try_future/mod.rs:
```diff
@@ -14,65 +14,121 @@ use futures_core::{
+    /// Future for the [`unwrap_or_else`](TryFutureExt::unwrap_or_else) method.
+    UnwrapOrElse<Fut, F>(
+        Map<IntoFuture<Fut>, UnwrapOrElseFn<F>>
+    ): Debug + Future + FusedFuture + Sink + Stream + FusedStream + New[|x: Fut, f: F| Map::new(IntoFuture::new(x), unwrap_or_else_fn(f))]
```
```suggestion
    ): Debug + Future + FusedFuture + New[|x: Fut, f: F| Map::new(IntoFuture::new(x), unwrap_or_else_fn(f))]
```

taiki-e commented:
Hmm... CI says that the behavior of `TryFutureExt::{and_then, or_else}` has changed: https://travis-ci.com/github/rust-lang/futures-rs/jobs/323173386#L340-L345

taiki-e commented:
Ok, the error is due to the old future is no longer dropped before the closure is run.
This is what `TryChain` did (https://github.com/rust-lang/futures-rs/pull/1069/files#r200509088), but `TryFlatten` does not handle closure, so this needs to be done in the `Map`.

Specifically, it needs to change the `future` field of `Map` to `Option<Fut>` and replace it with `None` before executing the closure.

```rust
    #[project]
    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<T> {
        #[project]
        let Map { mut future, f } = self.project();
        let fut = future.as_mut().as_pin_mut().expect("Map polled after completion");
        let output = ready!(fut.poll(cx));
        future.set(None); // Drop the old futures before the closure runs.

        let f = f.take().unwrap();
        Poll::Ready(f.call_once(output))
    }
```

Diggsey commented:
@taiki-e Yeah, I spotted that issue on the last CI run. I would like to avoid increasing the size of `Map`, so unfortunately it looks like I'd have to reintroduce a `Chain`-style abstraction for these combinators. However, implementing `Chain` would be easier if `pin_project` were to support the `project_replace` method we discussed previously, that way I could avoid an extra Option.

I had a look at what it would take to implement such a method in `pin_project` and it's surprisingly difficult - there are some awkward edge-cases around what happens if the destructor of a pinned field panics, so short term I will go with the Option approach, and open an issue in the unsafe-code-guidelines repo to see what the contract should be for panicking pinned destructors...

Diggsey commented:
Eh, I'm being stupid - `Map` already has an `Option` for the `FnOnce` field, so I can just get rid of that and turn the whole struct into an enum without increasing the size at all, and then everything will compose nicely with no special casing 😄. The only problem is that this again relies on `project_replace`, but luckily I think there's an alternative I can use for this specific case.

Diggsey commented:
@taiki-e I believe I have addressed all the issues and the tests are now passing.

I introduced some `unsafe` code in `Map`: this would be removable if `project_replace` functionality is added to pin_project, but is required for the moment. However, I have tested it against MIRI and it passes.

taiki-e commented on futures-util/src/future/future/flatten.rs:
```diff
@@ -1,56 +1,165 @@
+                    self.set(Flatten::Empty);
+                    break output;
+                },
+                Flatten::Empty => return Poll::Pending,
```
```suggestion
                Flatten::Empty => panic!("Flatten polled after completion"),
```

taiki-e commented:
Also, this PR adds the following APIs:

* `FutureExt::map_into`
* `future::try_maybe_done`
* `TryFutureExt::ok_into`
* `TryFutureExt::try_flatten`

I feel `map_into` and `ok_into` may not be needed, but I don't have a strong opinion on this, so I leave it to other reviewers.

cramertj commented:
This is outstanding work! I took a quick read through and I'm really excited to get this in. Thanks for all your hard work, @Diggsey, and for the detailed review, @taiki-e! There is quite a bit of code change here, so I'll do a release later this week to start getting this out to folks and make sure that there isn't any accidental breakage lurking.
